skbio/alignment/_path.py

```diff
@@ -0,0 +1,308 @@
+            # lens = np.diff(idx, append=ints.size)
+            lens = np.append(idx[1:] - idx[:-1], ints.size - idx[-1])
+
+            # Keep indices of segment starts.
+            ints = ints[idx]
+
+        # This is the 2D equivalent of the above code.
+        else:
+            idx = np.append(
+                0, np.where((ints[:, :-1] != ints[:, 1:]).any(axis=0))[0] + 1
+            )
+            lens = np.append(idx[1:] - idx[:-1], ints.shape[1] - idx[-1])
+            ints = ints[:, idx]
+
+        # return per-segment lengths and bits
+        return cls(lens, ints, np.zeros(bits.shape[0], dtype=int))
+
+    @classonlymethod
+    def from_tabular(cls, msa):
+        """Create an alignment path from a `TabularMSA` object."""
+        # Convert TabularMSA in to a 2D array of bytes.
+        # TODO: TabularMSA itself should be refactored to have this as the default data
+        # structure.
+        byte_arr = np.stack([x._bytes for x in msa._seqs])
+
+        # Get gap character code.
+        gap_chars = [ord(x) for x in msa.dtype.gap_chars]
+
+        # Identify gap positions, and convert them into a bit array, then create an
+        # alignment path based on it.
+        return cls.from_bits(np.isin(byte_arr, gap_chars))
+
+    def to_indices(self, gap=-1):
+        """Generate an array of indices of characters in the original sequences.
+
+        gap = "del": delete columns that have any gap, "mask": mask gaps, others
+        (default: -1):
+        fill gaps with this value.
+        """
+        valid_gaps = {-1, "del", "mask"}
+        if gap not in valid_gaps:
+            raise ValueError("Gap must be -1, 'del', or 'mask'.")
+
+        bits = self.to_bits()
+        # TODO: Consider optimization using np.arange.
+        # thought: initiate [-1, -1, -1 ... -1], then add slices of arange into it
+        pos = np.repeat(1 - bits, self.lengths, axis=1)
+        idx = np.cumsum(pos, axis=1, dtype=int) - 1
+        if gap == "del":
+            keep = np.repeat(self.states == 0, self.lengths)
+            return idx[:, keep]
+        elif gap == "mask":
+            return np.ma.array(idx, mask=(1 - pos))
+        else:
+            return np.where(pos, idx, gap)
+
+    @classonlymethod
+    def from_indices(cls, indices, gap=-1):
+        """Create an alignment path from character indices in the original sequences.
+
+        gap can be a value or "mask", but cannot be "del".
+        """
+        if gap == "mask":
+            return cls.from_bits(np.ma.getmask(indices))
+        else:
+            return cls.from_bits(indices == gap)
+
+    def to_coordinates(self):
+        """Generate an array of segment coordinates in the original sequences."""
+        lens = self.lengths * (1 - self.to_bits())
+        col0 = np.zeros((self.shape[0], 1), dtype=int)
+        lens = np.append(col0, lens, axis=1)
+        return lens.cumsum(axis=1)
+
+    @classonlymethod
+    def from_coordinates(cls, coords):
```

Comment:
```suggestion
    def from_coordinates(cls, coords):
        r"""Generate the an alignment path from an array of segment coordinates."""
```
Don't forget the docstring, this won't be documented otherwise

Reply:
Also need to document the parameteres